enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+For the best performance XDP programs need to attach to interfaces using the following models:
+- Native XDP
+  The XDP program is loaded by the network card driver as part of its initial receive path. This also requires support from the network card driver.
+- Offloaded XDP
+  The XDP program loads directly on the NIC, and executes without using the CPU. This requires support from the network interface device.
+
+## Motivation
+
+Customers would like to secure OCP cluster from external attacks.
+
+### Goals
+
+- flexible match filters rules for customer to use with a CRD validation to detect and fail invalid configuration.
+- Provide a mechanism for customers to select which nodes via `nodeSelector`.
+- Provide a mechanism to define attachement interfaces for the node ingress firewall objects.  
+- Ingress node firewall is deployed on a fully operational OCP cluster (i.e. cluster comes up with no default ingress firewall policy).
+- fail safe in case of user provisioned invalid rules which can lockup cluster operation.
+
+### Non-Goals
+- Flood Attacks
+
+## Proposal
+
+- customers will provision `NodeEndPoint` objects where the nodeFirewall policy will be applied to, if the list is empty the policy will be applied on all interfaces.
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will prepare `IngressNodeFirewall` objects with all required firewall rules and deploy, and via annotation will select the `NodeEndPoint` where the policy will be applied to, if there is no annotation rules will be attached to all interfaces.
+- IngressNodeFW controller will process `IngressNodeFirewall` objects, making sure configuration doesn't contradict with fail safe rules, update eBPF ingress table and attach eBPF kernel hook loaded as `BPF_PROG_TYPE_XDP` to the list of interfaces configured in `NodeEndPoint` object.
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
+
+### Workflow Description
+
+- the following diagram shows how node ingress firewall configuration is handled
+
+![Ingress Firewall configuration processing](./ingress-nodefw-config-flow.png)
+
+- the following diagram shows eBPF how ingress firewall kernel hook will process packets and apply firewall rules
+
+![eBPF workflow](./ebpf-packet-flow.png)
+
+- the following table shows sample of fail safe rules to make sure users won't block basic cluster services by configuring invalid firewall rules
+
+| Source CIDR | Protocol | Port | Purpose |
+|-------------|----------|------|---------|
+|  0.0.0.0/0  |   TCP    |  22  |  SSH    |
+|  0.0.0.0/0  |   UDP    |  68  | DHCP    |
+|  0.0.0.0/0  |   TCP    | 6443 | Kubernetes API server access |
+|  0.0.0.0/0  |   TCP    | 2380 & 2379 | Etcd access |
```

Comment:
What about the network plugin's ports? (such as for geneve, vxlan). Should the failsafe include communication to/from other cluster nodes?

Reply:
Again are these handled internally or managed by OCP?

Reply:
failsafe rules only cover criticial cluster operations TCP and UDP ports like API server, etcd , any other type of ports we assume admin need to know what he is doing or we can easily extend our rules list to cover any critical ports as needed